File: python/Minimum_Cost_Path_with_Edge_Reversals_3650.py
class Solution:
    def minCost(self, n: int, edges: List[List[int]]) -> int:
        edgeList, nodeDepd = [[] for _ in range(n)], [[] for _ in range(n)]
        for edge in edges:
            edgeList[edge[0]].append([edge[1], edge[2]])
            nodeDepd[edge[1]].append([edge[0], edge[2]])
        nodeLens = [1e8 for _ in range(n)]
        nextEle = [(0, 0)]
        while nextEle:
            curDist, curNode = heapq.heappop(nextEle)
            if curDist > nodeLens[curNode]:
                continue
            for node, edgeLen in edgeList[curNode]:
                if nodeLens[node] > curDist + edgeLen:
                    nodeLens[node] = curDist + edgeLen
                    heapq.heappush(nextEle, (nodeLens[node], node))
            for node, edgeLen in nodeDepd[curNode]:
                if nodeLens[node] > curDist + 2 * edgeLen:
                    nodeLens[node] = curDist + 2 * edgeLen
                    heapq.heappush(nextEle, (nodeLens[node], node))
        return -1 if nodeLens[-1] is 1e8 else nodeLens[-1]

File: cpp/Minimum_Cost_Path_with_Edge_Reversals_3650.cpp
class Solution {
public:

    static constexpr int unVisited = 1e8;

    int minCost(int n, vector<vector<int>>& edges) {
        vector <vector <pair <int, int>>> edgeList(n), nodeDepd(n);
        for (auto& edge: edges) {
            edgeList[edge[0]].push_back(make_pair(edge[1], edge[2]));
            nodeDepd[edge[1]].push_back(make_pair(edge[0], edge[2]));
        }
        vector <int> nodeLens(n, unVisited);
        nodeLens[0] = 0;
        priority_queue <tuple <int, int>, vector <tuple<int, int>>, greater<tuple<int, int>>> nextEle;
        nextEle.emplace(0, 0);
        while (!nextEle.empty()) {
            auto [curDist, curNode] = nextEle.top();
            nextEle.pop();
            if (nodeLens[curNode] < curDist) continue;
            for (auto&[node, edgeLen]: edgeList[curNode]) {
                if (nodeLens[node] > curDist + edgeLen) {
                    nodeLens[node] = curDist + edgeLen;
                    nextEle.emplace(nodeLens[node], node);
                }
            }
            for (auto&[node, edgeLen]: nodeDepd[curNode]) {
                if (nodeLens[node] > curDist + 2 * edgeLen) {
                    nodeLens[node] = curDist + 2 * edgeLen;
                    nextEle.emplace(nodeLens[node], node);
                }
            }
        }
        return nodeLens.back() == unVisited ? -1 : nodeLens.back();
    }
};
